tools/backups2datalad/zarr.py

```diff
@@ -0,0 +1,405 @@
+                        key,
+                        f"{self.api_url}/zarr/{self.zarr_id}.zarr/{entry}",
+                    )
+                    if (
+                        remotes is not None
+                        and self.backup_remote is not None
+                        and self.backup_remote not in remotes
+                    ):
+                        log.info(
+                            "Zarr %s: %s: Not in backup remote %s",
+                            self.zarr_id,
+                            entry,
+                            self.backup_remote,
+                        )
+        old_checksum: Optional[str]
+        try:
+            old_checksum = (self.repo / CHECKSUM_FILE).read_text().strip()
+        except FileNotFoundError:
+            old_checksum = None
+        if old_checksum != self.checksum:
+            log.info("Zarr %s: Updating checksum file", self.zarr_id)
+            (self.repo / CHECKSUM_FILE).write_text(f"{self.checksum}\n")
+            self.report.checksum = True
+        await trio.to_thread.run_sync(self.prune_deleted)
+
+    def rmtree(self, dirpath: Path) -> None:
+        for p in list(dirpath.iterdir()):
+            if p.is_dir():
+                self.rmtree(p)
+            else:
+                log.info("Zarr %s: deleting %s", self.zarr_id, p)
+                p.unlink()
+                self.report.deleted += 1
+        dirpath.rmdir()
+
+    def prune_deleted(self) -> None:
+        log.info("Zarr %s: deleting extra files", self.zarr_id)
+        dirs = deque([self.repo])
+        empty_dirs: deque[Path] = deque()
+        while dirs:
+            d = dirs.popleft()
+            is_empty = True
+            for p in list(d.iterdir()):
+                path = p.relative_to(self.repo).as_posix()
+                if d == self.repo and p.name in (
+                    CHECKSUM_FILE,
+                    ".datalad",
+                    ".git",
+                    ".gitattributes",
+                ):
+                    pass
+                elif p.is_dir():
+                    dirs.append(p)
+                elif path not in self.extant_paths:
+                    log.info("Zarr %s: deleting %s", self.zarr_id, path)
+                    p.unlink()
+                    self.report.deleted += 1
+                else:
+                    is_empty = False
+            if is_empty and d != self.repo:
+                empty_dirs.append(d)
+        while empty_dirs:
+            d = empty_dirs.popleft()
+            d.rmdir()
+            if d.parent != self.repo and not any(d.parent.iterdir()):
+                empty_dirs.append(d.parent)
+
+    async def aiter_file_entries(self) -> AsyncIterator[RemoteZarrEntry]:
+        root = RemoteZarrEntry(client=None, zarr_id=self.zarr_id, parts=())
+        dirs = deque([root])
+        async with httpx.AsyncClient() as client:
+            while dirs:
+                d = dirs.popleft()
+                path = "".join(p + "/" for p in d.parts)
```

Comment:
curious -- may be
```suggestion
                path = str(d).rstrip('/') + '/'
```
if the goal is just to ensure trailing `/` , or am I missing some aspect (e.g. could be empty?)

Reply:
Yes, `path` needs to be empty when `d` is `root`.